receiver/sqlqueryreceiver/config.go

```diff
@@ -28,16 +31,126 @@ type Config struct {
+	}
+	for _, query := range c.Queries {
+		if err := query.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 type Query struct {
-	SQL     string   `mapstructure:"sql"`
-	Metrics []Metric `mapstructure:"metrics"`
+	SQL     string      `mapstructure:"sql"`
+	Metrics []MetricCfg `mapstructure:"metrics"`
 }
 
-type Metric struct {
-	MetricName       string   `mapstructure:"metric_name"`
-	ValueColumn      string   `mapstructure:"value_column"`
-	AttributeColumns []string `mapstructure:"attribute_columns"`
-	IsMonotonic      bool     `mapstructure:"is_monotonic"`
+func (q Query) Validate() error {
+	var errs error
+	if q.SQL == "" {
+		errs = multierr.Append(errs, errors.New("'query.sql' cannot be empty"))
+	}
+	if len(q.Metrics) == 0 {
+		errs = multierr.Append(errs, errors.New("'query.metrics' cannot be empty"))
+	}
+	for _, metric := range q.Metrics {
+		if err := metric.Validate(); err != nil {
+			errs = multierr.Append(errs, err)
+		}
+	}
+	return errs
+}
+
+type MetricCfg struct {
+	MetricName       string            `mapstructure:"metric_name"`
+	ValueColumn      string            `mapstructure:"value_column"`
+	AttributeColumns []string          `mapstructure:"attribute_columns"`
+	Monotonic        bool              `mapstructure:"monotonic"`
+	ValueType        MetricValueType   `mapstructure:"value_type"`
+	DataType         MetricDataType    `mapstructure:"data_type"`
+	Aggregation      MetricAggregation `mapstructure:"aggregation"`
+	Unit             string            `mapstructure:"unit"`
+	Description      string            `mapstructure:"description"`
+}
+
+func (c MetricCfg) Validate() error {
+	var errs error
+	if c.MetricName == "" {
+		errs = multierr.Append(errs, errors.New("metric config: 'metric_name' cannot be empty"))
+	}
+	if c.ValueColumn == "" {
+		errs = multierr.Append(errs, errors.New("metric config: 'value_column' cannot be empty"))
+	}
+	if err := c.ValueType.Validate(); err != nil {
+		errs = multierr.Append(errs, err)
+	}
+	if err := c.DataType.Validate(); err != nil {
+		errs = multierr.Append(errs, err)
+	}
+	if err := c.Aggregation.Validate(); err != nil {
+		errs = multierr.Append(errs, err)
+	}
+	if errs != nil && c.MetricName != "" {
+		errs = multierr.Append(fmt.Errorf("invalid metric config with metric_name '%s'", c.MetricName), errs)
+	}
+	return errs
+}
+
+type MetricDataType string
+
+const (
+	MetricDataTypeUnspecified MetricDataType = ""
+	MetricDataTypeGauge       MetricDataType = "gauge"
+	MetricDataTypeSum         MetricDataType = "sum"
+)
+
+func (t MetricDataType) Validate() error {
+	switch t {
+	case MetricDataTypeUnspecified, MetricDataTypeGauge, MetricDataTypeSum:
+		return nil
+	}
+	return fmt.Errorf("metric config has unsupported data_type: '%s'", t)
+}
+
+type MetricValueType string
+
+const (
+	MetricValueTypeUnspecified MetricValueType = ""
+	MetricValueTypeInt         MetricValueType = "int"
+	MetricValueTypeDouble      MetricValueType = "double"
```

Comment:
Sorry, another late comment. The `README` says `float`, but this says `double`. I don't necessarily have a preference either way, it should just be consistent.

Reply:
Good catch, fixed. Going with float.